Improvements of metrics gathering for RestTemplate calls

ArturGajowy opened the pull request:
- introduced 'rest.client.metrics.enabled' property (enabled by default) for toggling gathering of metrics for requests made with RestOperations instances
- renamed MetricsAspect to RestOperationsMetricsAspect,
- introduced UrlMetricNamer - a configurable collaborator Strategy of RestOperationsMetricsAspect used to determine a sensible (not containing dynamic parts) metric name for any URL used in a rest call
- added RegexMatchingPathElidingURIMetricNamer as a default implementation,
- Made RestOperationsMetricsAspect correctly intercept all methods of RestOperations, added integration test
- added Utility methods for getting wireMock properties in MvcWiremockIntegrationSpec

kamilszymanski commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/UrlMetricNameProvider.groovy:
```diff
@@ -0,0 +1,63 @@
+import static org.springframework.util.StringUtils.trimLeadingCharacter
+import static org.springframework.util.StringUtils.trimTrailingCharacter
+
+@CompileStatic
```
As you know we're in the middle of rewriting micro-infra-spring to Java.
Could you rewrite `UrlMetricNameProvider` to Java

kamilszymanski commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/UrlMetricNameProvider.groovy:
```diff
@@ -0,0 +1,63 @@
+import static org.springframework.util.StringUtils.trimTrailingCharacter
+
+@CompileStatic
+class UrlMetricNameProvider {
```
Please provide javadocs for `UrlMetricNameProvider`

kamilszymanski commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/MetricsAspect.groovy:
```diff
@@ -18,42 +17,43 @@ class MetricsAspect {
+            String approximateUriString = urlTemplateString.replaceAll(/[{}]/, '')
+            uri = new URI(approximateUriString)
+        } else {
+            throw new IllegalArgumentException(
```
Instrumentation code should not make valid operations fail

kamilszymanski commented on micro-infra-spring-base/src/test/groovy/com/ofg/infrastructure/web/resttemplate/MetricsAspectIntegrationSpec.groovy:
```diff
@@ -0,0 +1,143 @@
+            verifyMetric('.key') { restTemplate.delete("$wireMockUrl/{key}", [key: 'value']) }
+            verifyMetric         { restTemplate.delete(wireMockUri) }
+
+            def responseType = Void
```
Given part in expect block suggests that this should be another (appropriately named) test

ArturGajowy commented:
Added javadocs, made the thing log.warn instead of throwing an exception in the rare event Spring changes the signatures of RestOperations' methods, refactored the tests as suggested by @kamilszymanski.

ArturGajowy commented:
got:
```
FAILURE: Build failed with an exception.
* What went wrong:
Could not resolve all dependencies for configuration ':testCompile'.
> Could not find com.jayway.restassured:rest-assured:2.4.1microInfraSpringVersion=0.8.25-SNAPSHOT.
```
for both failing builds in jdk 7 jobs. The dependency
`com.jayway.restassured:rest-assured:2.4.1microInfraSpringVersion=0.8.25-SNAPSHOT`
looks like an error in build.gradle or other build infrastructure. Couldn't find it though. Does not seem to be related with my changes, as I didn't change any of those. Any ideas?

ArturGajowy commented:
K, found it, all is well now. Can I have my +1? ;)

szpak commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/RestOperationsMetricsAspect.groovy:
```diff
@@ -0,0 +1,58 @@
+    private URI getUriForMetricNaming(ProceedingJoinPoint pjp) {
+        def firstArgument = pjp.args[0]
+        URI uri
+        if (firstArgument instanceof URI) {
```
A matter of taste, but probably switch..case could be cleaner.

ArturGajowy replied:
I really like to have one return statement, at the end of the method - that's why no switch. I could use break, but that would defeat the purpose of using switch, I think. Also, I would not like to introduce too much groovyness when we're migrating to Java, so no `case URI`, `case String` :)

szpak commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/RestOperationsMetricsAspect.groovy:
```diff
@@ -0,0 +1,58 @@
+            def requestEntity = (RequestEntity) firstArgument
+            uri = requestEntity.url
+        } else if (firstArgument instanceof String) {
+            String urlTemplateString = (String) firstArgument
```
Is it still required to cast it in Groovy (after `if (... instanceof ...)`)?

ArturGajowy replied:
It seemed to me it was, which occurred strange to me at that moment as well... But again, this will save someone some work with Java 7 migration.

szpak commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/ServiceRestClientConfigurationSupport.groovy:
```diff
@@ -120,8 +122,17 @@ class ServiceRestClientConfigurationSupport {
+    @ConditionalOnExpression('${rest.client.metrics.enabled:true}')
+    RestOperationsMetricsAspect restOperationsMetricsAspect(
+            MetricRegistry metricRegistry, URIMetricNamer uriMetricNamer
+    ) {
```
Could be in the previous line

ArturGajowy replied:
I'd cross the 120 chars limit ;)

szpak replied:
Just the brackes `) {`.

szpak commented on micro-infra-spring-base/src/test/groovy/com/ofg/infrastructure/web/resttemplate/MetricsAspectIntegrationSpec.groovy:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    def 'should elide subsequent path parts in metric name'() {
+        expect:
```
Couldn't `where` be used here?

ArturGajowy replied:
I tried really hard to do that, but then every closure would have 2 parameters. where-blocks can only access static scope and I need the restTemplate and wireMockUrl instance fields, so I'd prefer it this way. Why would you like a where-block? To have failures reported separately? Then this test would count for like 40 tests, which wouldn't be fair I think ;).

szpak commented on micro-infra-spring-base/src/test/groovy/com/ofg/infrastructure/web/resttemplate/MetricsAspectIntegrationSpec.groovy:
```diff
@@ -0,0 +1,154 @@
+        MetricRegistry metricRegistry() {
+            def lastMetricName = 'UNSET'
+            return [
+                    timer              : { String metricName ->
```
Strange formatting

ArturGajowy replied:
you mean the spaces after the timer or the missalignment with the last colon (line 145)?

szpak replied:
Spaces before the colon.

szpak commented on gradle/runBootMicroserviceAcceptanceTests.sh:
```diff
@@ -7,8 +7,7 @@ function dumpCurrentMicroInfraSpringVersionToFile {
 function updateMicroInfraSpringVersionInConfigurationFile {
-
-    echo "microInfraSpringVersion="`cat ~/.microInfraSpringCurrentVersion.txt` >> gradle.properties
+    echo -e "\nmicroInfraSpringVersion="`cat ~/.microInfraSpringCurrentVersion.txt` >> gradle.properties
```
Good catch

ArturGajowy replied:
It was Travis that caught it ;)

szpak commented on micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/ServiceRestClientConfigurationSupport.groovy:
```diff
@@ -120,8 +122,17 @@ class ServiceRestClientConfigurationSupport {
+    }
+
+    @Bean
+    @ConditionalOnMissingBean(URIMetricNamer)
```
Have you tested it manually in the "real" application?

ArturGajowy replied:
Just tested it with boot-microservice and it works as expected. Why would we test that manually? It's Spring's feature after all...

szpak replied:
This AFAIR is supported (guaranteed) only when used in the AutoConfiguration mechanism. But if it worked for you (you were able to override it) then we will be worried about on the first (potential) production failure.

szpak commented:
Have you run Cola or something else with your changes?

ArturGajowy commented:
@szpak I've done the formatting changes you wanted. Can we merge? :)

kamilszymanski commented:
As you know we're in the middle of rewriting micro-infra-spring to Java.
Please rewrite newly introduced Groovy classes to Java.

ArturGajowy commented:
@kamilszymanski done. Please note RestOperationsMetricsAspect.groovy is renamed from MetricsAspect.groovy

kamilszymanski commented:
+1 if you've tested it against service that does not run on spring-boot
